src/gatsby-node.js

```diff
@@ -0,0 +1,577 @@
+          },
+          {
+            name: "memory-heapTotal",
+            type: "gauge",
+            value: heapTotal !== null && heapTotal !== void 0 ? heapTotal : 0,
+            timestamp: timestamp,
+            attributes: attributes,
+          },
+          {
+            name: "memory-heapUsed",
+            type: "gauge",
+            value: heapUsed !== null && heapUsed !== void 0 ? heapUsed : 0,
+            timestamp: timestamp,
+            attributes: attributes,
+          },
+          {
+            name: "memory-external",
+            type: "gauge",
+            value: external !== null && external !== void 0 ? external : 0,
+            timestamp: timestamp,
+            attributes: attributes,
+          },
+          {
+            name: "build-times",
+            type: "gauge",
+            value: timeelapsed,
+            timestamp: timestamp,
+            attributes: buildtimes,
+          },
+        ],
+      },
+    ];
+  }
+
+  markStart() {
+    if (this.started) {
+      reportError(
+        `[@] gatsby-plugin-newrelic: `,
+        new Error(`Error: Should not call markStart() more than once`)
+      );
+      process.exit(1);
+    }
+
+    this.timestamps.benchmarkStart = performance.now();
+    this.started = true;
+  }
+
+  markDataPoint(name) {
+    if (BENCHMARK_REPORTING_URL) {
+      if (!(name in this.timestamps)) {
+        reportError(
+          `[@] gatsby-plugin-newrelic: Attempted to record a timestamp with a name (\`${name}\`) that wasn't expected`
+        );
+        process.exit(1);
+      }
+    }
+
+    this.timestamps[name] = performance.now();
+  }
+
+  async markEnd() {
+    if (!this.timestamps.benchmarkStart) {
+      reportError(
+        `[@] gatsby-plugin-newrelic:`,
+        new Error(`Error: Should not call markEnd() before calling markStart()`)
+      );
+      process.exit(1);
+    }
+
+    this.timestamps.benchmarkEnd = performance.now();
+    return this.flush();
+  }
+
+  async flush() {
+    const data = this.getData();
+    const json = JSON.stringify(data, null, 2);
+
+    if (!BENCHMARK_REPORTING_URL) {
+      reportInfo(
+        `[@] gatsby-plugin-newrelic: MetricAPI BENCHMARK_REPORTING_URL not set, not submitting data`
+      );
+      this.flushed = true;
+      return (this.flushing = Promise.resolve());
+    }
+    if (!NR_INGEST_KEY) {
+      console.log(`[!] gatsby-plugin-newrelic: NR_INGEST_KEY not set`);
+      this.flushed = true;
+      return (this.flushing = Promise.resolve());
+    }
+
+    reportInfo(
+      `[@] gatsby-plugin-newrelic: Flushing benchmark data to remote server...`
+    );
+    let lastStatus = 0;
+    this.flushing = nodeFetch(`${BENCHMARK_REPORTING_URL}`, {
+      method: `POST`,
+      headers: {
+        "content-type": `application/json`,
+        "Api-Key": NR_INGEST_KEY,
+      },
+      body: json,
+    }).then((res) => {
+      lastStatus = res.status;
+
+      if ([401, 500].includes(lastStatus)) {
```

Comment:
Wouldn't we _not_ want to see a `500` response code? I wonder if we're suppose to `throw` / `catch` in the event that a server error is returned.

Reply:
Ah I see, this _is_ for cases when we are handling the error state. We still might want to `throw` and `catch` down below, since that's more descriptive of what we're doing.

Reply:
decided not to use `throw` here as the try/catch is only used in the failsafe flush within `process.on(exit`